### PR TITLE
[SPARK-40721][CORE] Use `||` instead of `|` in `TimSort`

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This pr changes `while (count1 >= MIN_GALLOP | count2 >= MIN_GALLOP)` in `TimSort` to `while (count1 >= MIN_GALLOP || count2 >= MIN_GALLOP)`.

### Why are the changes needed?
The or operation of 2 boolean expressions should use `||`.
When `||` is used, if the first expression is true, the second expression does't need to be calculated. When `|` is used, 2 expressions must be calculated.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass GitHub Actions

### CODE REVIEW

File: core/src/main/java/org/apache/spark/util/collection/TimSort.java [hunk 1]
```diff
@@ -788,7 +788,7 @@ private void mergeLo(int base1, int len1, int base2, int len2) {
           if (--len1 == 1)
             break outer;
           minGallop--;
-        } while (count1 >= MIN_GALLOP | count2 >= MIN_GALLOP);
+        } while (count1 >= MIN_GALLOP || count2 >= MIN_GALLOP);
```
Comment (dongjoon-hyun) [inline core/src/main/java/org/apache/spark/util/collection/TimSort.java:791]: Thank you. Actually, this was a copy of Android TimSort.java. I'm just curious if Android change like your proposal recently?
- https://android.googlesource.com/platform/libcore/+/android-6.0.1_r77/luni/src/main/java/java/util/TimSort.java#749

**Comment by mridulm** [comment]
There is a reason to use `|` instead of `||` - you reduce chances of branch misprediction and minimize number of jumps - if the operations are really cheap.
Having said that, I have not inspected the opcodes generated here - but I would assume that should hold.

Note - this is typically used only for very performance sensitive code, where someone has gone to the trouble of actually ensuring the optimization actually gives the expected benefits.

File: core/src/main/java/org/apache/spark/util/collection/TimSort.java (hunk 1 above)
Comment (mridulm) [reply core/src/main/java/org/apache/spark/util/collection/TimSort.java:791]: Is the same in master: https://android.googlesource.com/platform/libcore/+/refs/heads/master/ojluni/src/main/java/java/util/TimSort.java#769

**Comment by LuciferYang** [comment]
Thanks for your answers @mridulm @dongjoon-hyun , I will close this one

**Comment by dongjoon-hyun** [comment]
Thank you, @mridulm and @LuciferYang .

**Comment by srowen** [comment]
Funny story, I think I remember discussing this exact thing with Josh Bloch (of "Effective Java") maaany years ago and he confirmed this is used on purpose in tight loops in some cases because it's actually faster. I suspect this is indeed on purpose from the original TimSort code

**Comment by LuciferYang** [comment]
> Funny story, I think I remember discussing this exact thing with Josh Bloch (of "Effective Java") maaany years ago and he confirmed this is used on purpose in tight loops in some cases because it's actually faster. I suspect this is indeed on purpose from the original TimSort code

Interesting, every rule has its exception, seems this is the exception :)

**Pull request closed without merging**